File: services/product-service/src/main/java/com/zerofake/product/service/ProductService.java
package com.zerofake.product.service;

import com.zerofake.product.dto.request.CreateProductRequest;
import com.zerofake.product.dto.request.UpdateProductRequest;
import com.zerofake.product.dto.common.ApiResponse;
import com.zerofake.product.dto.response.ProductResponse;

import java.util.List;
import java.util.UUID;

public interface ProductService {

    ProductResponse createProduct(CreateProductRequest request);

    List<ProductResponse> getAllProducts();

    ProductResponse getProductById(UUID id);

    ProductResponse updateProduct(UUID id, UpdateProductRequest request);

    ApiResponse<Void> deleteProduct(UUID id);

    List<ProductResponse> getProductsByCategory(UUID categoryId);

    List<ProductResponse> getProductsByManufacturer(UUID manufacturerId);

}

File: zerofake-frontend/src/services/productService.ts
import { productApi } from "@/api/productApi";

import type {
  ProductResponse,
  CreateProductRequest,
  UpdateProductRequest,
} from "@/types/product";

class ProductService {
  async getProducts(): Promise<ProductResponse[]> {
    const response =
      await productApi.getAllProducts();

    return response.data ?? [];
  }

  async refreshProducts(): Promise<ProductResponse[]> {
    return this.getProducts();
  }

  async createProduct(
    request: CreateProductRequest
  ): Promise<ProductResponse> {
    const response =
      await productApi.createProduct(request);

    return response.data;
  }
  async updateProduct(
  productId: string,
  request: UpdateProductRequest
): Promise<ProductResponse> {
  const response =
    await productApi.updateProduct(
      productId,
      request
    );

  return response.data;
}
  async deleteProduct(
    productId: string
  ): Promise<void> {
    await productApi.deleteProduct(productId);
  }

  async getProduct(productId: string): Promise<ProductResponse> {
    const response = await productApi.getProductById(productId);
    return response.data;
  }

  async updateBlockchainStatus(productId: string, status: string): Promise<ProductResponse> {
    const response = await productApi.updateBlockchainStatus(productId, status);
    return response.data;
  }
}

export const productService =
  new ProductService();